numpy/core/function_base.py

```diff
@@ -1,10 +1,24 @@
 from __future__ import division, absolute_import, print_function
 
+import warnings
+import operator
+
 __all__ = ['logspace', 'linspace']
 
 from . import numeric as _nx
 from .numeric import result_type, NaN, shares_memory, MAY_SHARE_BOUNDS, TooHardError
 
+def _index_deprecate(i, stacklevel=2):
```

Comment:
Hmm, try/except seems simple and obviously correct to me. I think checking for `__index__` is probably correct but am not as confident, and I'm pretty sure checking for `Integral` is subtly wrong. So I'd vote for the current implementation, on the ground that we should only use style as a tie breaker when we're version that all the options being considered are correct :-).

It doesn't much matter whether we have a helper function here or write the same code inline, but I think the helper function got added at someone (@seberg's?) request in an earlier round of review, and I do agree that it makes things easier to read. I think the original motivation was that we have a similar helper at the C level and it's useful to use in multiple places; here we only have one user so it doesn't matter as much, but potentially there will be other users in the future.